torch/_inductor/runtime/benchmarking.py

```diff
@@ -0,0 +1,944 @@
+        # on the CPU; this is bad since we try to overlap the CPU launch period with a
+        # GPU sleep to avoid the CPU launch overhead in our GPU event timings. what ends
+        # up happening in this case is that the self.gpu_queue_limit'th event will wait
+        # until the queue shrinks (i.e. after the GPU sleep ends) before queueing itself.
+        # this has two issues, first we will waste a lot of time needlesly as the CPU
+        # waits for the GPU to finish its sleep, and second we will no longer have a
+        # proper overlap between the CPU launch overhead and the GPU sleep which can
+        # result in incorrect timing values for particularly small kernels as the CPU
+        # will eventually catch up to the GPU and we will end up in lock-step execution.
+        # we can prevent this by benchmarking in blocks. we calculate the number of blocks
+        # required by assuming that each benchmark iteration spawns approximately 5 events:
+        # 1 for the buffer zeroing, 1 for the callable, 2 for the event records, and 1
+        # extra allocated for any sub-kernels the callable may launch of which the most
+        # common case is allocating an output Tensor. we then evenly split the total
+        # number of benchmark iterations across all blocks.
+        benchmark_blocks = ((benchmark_iters * 5) // (self.gpu_queue_limit - 1)) + 1
+        benchmark_iters_per_block = benchmark_iters // benchmark_blocks
+        memory_warmup_iters_per_block = memory_warmup_iters // benchmark_blocks
+        required_gpu_sleep_cycles_per_block = self.get_required_gpu_sleep_cycles(
+            memory_warmup_iters_per_block,
+            benchmark_iters_per_block,
+            cpu_launch_overhead_ms_per_iter,
+        )
+
+        # ensures the queue is empty
+        torch.cuda.synchronize()
+        event_pairs = self.get_event_pairs(benchmark_iters)
+        for block_idx in range(benchmark_blocks):
+            block_start = block_idx * benchmark_iters_per_block
+            if block_idx == (benchmark_blocks - 1):
+                block_end = benchmark_iters
+            else:
+                block_end = (block_idx + 1) * benchmark_iters_per_block
+            torch.cuda._sleep(required_gpu_sleep_cycles_per_block)
+            for _ in range(memory_warmup_iters_per_block):
+                buffer.zero_()
+            for start_event, end_event in event_pairs[block_start:block_end]:
+                buffer.zero_()
+                start_event.record()
+                _callable()
+                end_event.record()
+            torch.cuda.synchronize()
+        timing_ms = self.get_min_timing_ms(event_pairs)
+
+        # explicitly delete the buffer, sometimes helps memory
+        # footprint metrics in OSS Inductor benchmarks
+        del buffer
+
+        # we may as well use the estimation loop signal as well
+        return min(estimated_timing_ms, timing_ms)
```

Comment:
I wold also call out in the docstring for this function that it will return the minimum time which is a divergence from the existing triton benchmarking. 

also, cc, @int3, @bertmaher  if you have any thoughts. I know there have been some discussions previously about benchmarking with different clock rates